src/lib.rs

```diff
@@ -472,29 +472,30 @@ pub fn create(value: &EtherValue, data: &[u8]) -> CreateResult {
 }
 
 /// Executes callDataCopy, but does not check for overflow.
-pub fn unsafe_calldata_copy(from: usize, length: usize) -> Vec<u8> {
-    let mut ret: Vec<u8> = unsafe_alloc_buffer(length);
-
+pub fn unsafe_calldata_copy(from: usize, length: usize, ret: &mut [u8]) {
     unsafe {
         native::ethereum_callDataCopy(ret.as_mut_ptr() as *const u32, from as u32, length as u32);
     }
-
-    ret
 }
 
 /// Returns a vector containing all data passed with the currently executing call.
 pub fn calldata_acquire() -> Vec<u8> {
-    unsafe_calldata_copy(0, calldata_size())
+    let length = calldata_size();
+
+    let mut ret: Vec<u8> = unsafe_alloc_buffer(length);
+    unsafe_calldata_copy(0, length, &mut ret);
+    ret
 }
 
 /// Returns the segment of call data beginning at `from`, and continuing for `length` bytes.
-pub fn calldata_copy(from: usize, length: usize) -> Result<Vec<u8>, Error> {
+pub fn calldata_copy(from: usize, length: usize, ret: &mut [u8]) -> Result<(), Error> {
     let size = calldata_size();
 
-    if (size < from) || ((size - from) < length) {
+    if (size < from) || ((size - from) < length || ret.len() < length) {
```

Comment:
Should this require an exact slice?

Reply:
not sure what you mean. all slices have an associated fixed length

Reply:
Options
a) `ret.len() == length` or
b) good as long as the slice is bigger than what we want to copy (current implementation is this)

Reply:
If we have separate `calldata_acquire` method then a) is the obvious option. the question is what use cases a) would have, and whether they would be very ergonomic at this layer of abstraction.

Reply:
Ah right this question does needs an answer. `ret.len() == length` or `ret.len() >= length` should be the condition.

Reply:
second one would be unsafe if the same buffer is used to write multiple times, because underflow is possible.

Reply:
Hm? These are slices, there's no overflow. The assertion is `ret.len() <= length` or `ret.len() == length`.